### PR TITLE
Support for custom device capabilities

### PR DESCRIPTION
**Pull request by tamirOK**

### CODE REVIEW

File: aonvif/client.py [hunk 1]
```diff
@@ -245,6 +245,15 @@ class ONVIFCamera:
                         solution, this should only be used in "safe" environments.
                         Also, this cannot be used on AXIS camera, as every request is
                         authenticated, contrary to ONVIF standard.
+    :param capabilities: Dictionary of camera capabilities. It must have the following format:
+                         {
+                            "{capability_name_1}": {
```
Comment (martyanov) [inline aonvif/client.py]: Please provide real capability examples.

File: aonvif/client.py [hunk 2]
```diff
@@ -274,6 +284,7 @@ def __init__(
 
         # Currently initialized services
         self._services = {}
+        self._capabilities = capabilities
```
Comment (martyanov) [inline aonvif/client.py]: Maybe check for xaddrs keys?

Comment (martyanov) [reply aonvif/client.py]: If capabilities were provided statically.

File: aonvif/exceptions.py [hunk 3]
```diff
@@ -4,3 +4,7 @@
 
 class ONVIFError(Exception):
     """Base type for all ONVIF errors."""
+
+
+class InvalidCapabilitiesError(Exception):
```
Comment (martyanov) [inline aonvif/exceptions.py]: Broken exception hierarchy. Probably there is no reason to use the separate exception. If wrong capabilities lead to broken client behavior then runtime exception is more appropriate.

File: aonvif/client.py [hunk 4]
```diff
@@ -274,9 +284,34 @@ def __init__(
...
+        self._validate_capabilities()
 
         self.to_dict = ONVIFService.to_dict
 
+    def _validate_capabilities(self):
+        if self._capabilities is None:
```
Comment (martyanov) [inline aonvif/client.py]: Validation after attribute assignment looks weird. I think validation should go before assignment.

Comment (tamirOK) [reply aonvif/client.py]: Fixed

File: aonvif/exceptions.py (hunk 3 above)
Comment (tamirOK) [reply aonvif/exceptions.py]: Replaced exception with RuntimeError

File: aonvif/client.py [hunk 5]
```diff
@@ -274,9 +274,42 @@ def __init__(
...
+            raise RuntimeError('Capabilities type must be dictionary')
+
+        for key, capability in capabilities.items():
+            if not isinstance(key, str):
+                raise RuntimeError('Capabilities key type must be string')
+
+            if not isinstance(capability, dict):
+                raise RuntimeError('Capability type must be dictionary')
+
+            if not isinstance(capability.get('XAddr'), str):
```
Comment (martyanov) [inline aonvif/client.py]: If we only support `XAddr` key probably its presence should be explicitly verfied, the current message is sort of misleading, if `XAddr` was not provided at all.

Comment (tamirOK) [reply aonvif/client.py]: Fixed

File: aonvif/client.py [hunk 6]
```diff
@@ -298,9 +336,12 @@ async def update_xaddrs(self):
...
         # Get XAddr of services on the device
         self._xaddrs = {}
-        capabilities = await devicemgmt.GetCapabilities({'Category': 'All'})
-        for name in capabilities:
-            capability = capabilities[name]
+
+        if not self._capabilities:
+            self._capabilities = await devicemgmt.GetCapabilities({'Category': 'All'})
+
+        for name in self._capabilities:
```
Comment (martyanov) [inline aonvif/client.py:343]: `for name, capability in self._capabilities.items()`?

File: aonvif/client.py [hunk 7]
```diff
@@ -274,9 +274,47 @@ def __init__(
...
+        """
+
+        self._validate_capabilities(capabilities)
+        self._capabilities = capabilities
+
+    def _validate_capabilities(self, capabilities: dict):
+        if not isinstance(capabilities, dict):
+            raise RuntimeError('Capabilities type must be dictionary')
+
+        for key, capability in capabilities.items():
```
Comment (martyanov) [inline aonvif/client.py:303]: `name, capability`?

File: aonvif/client.py (hunk 6 above)
Comment (tamirOK) [reply aonvif/client.py:343]: `AttributeError: Capabilities instance has no attribute 'items'`

Comment (martyanov) [reply aonvif/client.py:343]: Go it, thanks!

**Review by martyanov — APPROVED** [review]

**Review by martyanov — APPROVED** [review]

**Pull request merged**